quill/ecs/src/entity.rs

```diff
@@ -0,0 +1,25 @@
+#[derive(Copy, Clone, Debug, PartialEq, Eq, Hash)]
+pub struct EntityId {
+    index: u32,
+    generation: u32,
+}
+
+impl EntityId {
+    pub fn to_bits(self) -> u64 {
+        ((self.index as u64) << 32) | (self.generation as u64)
+    }
+
+    pub fn from_bits(bits: u64) -> Self {
+        let index = (bits >> 32) as u32;
+        let generation = bits as u32;
+        Self { index, generation }
+    }
+
+    pub fn index(self) -> u32 {
+        self.index
+    }
+
+    pub fn generation(self) -> u32 {
+        self.generation
```

Comment:
What is `generation` used for, can't quite seem to connect it.

Reply:
It's used to solve the ABA problem, since entity indices are reused. A newly created entity could have the same index as an old, despawned entity, so we use the generation to ensure old `Entity` handles don't refer to new entities.

Reply:
Another approach could also be to queue up insertions and deletions of components and entities. And when the `Query` is dropped, commit the queue to the `World`.

Reply:
That would not let us get rid of generations

Reply:
True an old `Entity` id could be pointing to a new `Entity`, should we allow end users to store long lived entity ids?

Reply:
Yes, storing long-lived entities is definitely needed. Feather uses it for the `ChunkEntities` data structure, for example.